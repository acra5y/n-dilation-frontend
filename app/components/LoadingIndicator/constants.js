export const MAX_SYMBOLS = 3;
export const INTERVAL = 1000;
export const MINIMUM_TIME_VISIBLE = INTERVAL * (1 / 2);
export const ANIMATION_TIME = 100;
