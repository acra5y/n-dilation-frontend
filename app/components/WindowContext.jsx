import React, { useContext } from "react";

const WindowContext = React.createContext();

export const useWindowContext = () => useContext(WindowContext);

export default WindowContext;
