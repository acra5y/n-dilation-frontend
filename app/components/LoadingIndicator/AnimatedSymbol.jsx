import React, { useEffect, useState } from "react";
import styled from "styled-components";

import { ANIMATION_TIME, INTERVAL, MINIMUM_TIME_VISIBLE } from "./constants";
import getRandomInt from "../../lib/getRandomInt";
import { useWindowContext } from "../WindowContext";

const StyledSymbol = styled.span.attrs(props => ({
    style: {
        flexBasis: `${props.basis}%`,
        margin: `calc(4rem * ${props.offsetTop}) 3px 3px`,
        opacity: props.opacity,
    },
}))`
    white-space: nowrap;
    transition: opacity ${ANIMATION_TIME}ms ease-out;
`;
StyledSymbol.displayName = "StyledSymbol";

const AnimatedSymbol = props => {
    const [opacity, setOpacity] = useState(0);

    const window = useWindowContext();

    useEffect(() => {
        const { setTimeout, clearTimeout } = window;
        let i, j;

        const clearance = 100;

        const timeBeforeFadeIn = props.getRandomInt(clearance);
        const timeVisible = Math.max(
            MINIMUM_TIME_VISIBLE,
            props.getRandomInt(INTERVAL - timeBeforeFadeIn - ANIMATION_TIME)
        );

        i = setTimeout(() => {
            j = setTimeout(() => setOpacity(0), timeVisible);
            setOpacity(1);
        }, timeBeforeFadeIn);

        return () => {
            clearTimeout(i);
            clearTimeout(j);
        };
    }, [props.children]);

    return (
        <StyledSymbol
            basis={props.basis}
            offsetTop={props.offsetTop}
            opacity={opacity}
        >
            {props.children}
        </StyledSymbol>
    );
};

AnimatedSymbol.displayName = "AnimatedSymbol";

AnimatedSymbol.defaultProps = {
    getRandomInt,
};

export default AnimatedSymbol;
