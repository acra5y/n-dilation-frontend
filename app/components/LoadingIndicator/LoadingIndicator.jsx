import React, { useEffect, useState, useRef } from "react";
import styled from "styled-components";

import { MAX_SYMBOLS, INTERVAL } from "./constants";
import getRandomInt from "../../lib/getRandomInt";
import { useWindowContext } from "../WindowContext";
import AnimatedSymbol from "./AnimatedSymbol";

const symbolSet = [
    "∑",
    "±",
    "∆",
    "≤",
    "≈",
    "√",
    "∫",
    "~",
    "µ",
    "∞",
    "…",
    "| T |",
    "×",
    "∏",
    "|| x ||",
    "sup",
    "<x,y>",
    "T*",
    "dim(T)",
    "∃",
    "∀",
    "Re(z)",
    "Im(z)",
    "α1,...,αk",
    "|ε|=1",
    "λ",
    "j→∞",
    "ζ",
];

const getRandomSymbols = (getRandomIntFromProps) => {
    const amountOfSymbols = getRandomIntFromProps(MAX_SYMBOLS) + 1;

    return Array.from(Array(amountOfSymbols)).map(
        () => symbolSet[getRandomIntFromProps(symbolSet.length - 1)]
    );
};

const LoadingIndicatorWrapper = styled.div`
    display: flex;
    flex-direction: row ${({ reverse }) => (reverse ? "-reverse" : "")};
    flex-wrap: nowrap;
    height: 5rem;
    width: 300px;
`;

const LoadingIndicator = (props) => {
    const getRandomIntRef = useRef((arg) => arg - 1);
    const [symbols, setSymbols] = useState(
        getRandomSymbols(getRandomIntRef.current)
    );
    const window = useWindowContext();

    useEffect(() => {
        const { setInterval, clearInterval } = window;
        getRandomIntRef.current = props.getRandomInt;
        const i = setInterval(
            () => setSymbols(getRandomSymbols(getRandomIntRef.current)),
            INTERVAL
        );

        return () => clearInterval(i);
    }, []);

    return (
        <LoadingIndicatorWrapper reverse={Boolean(getRandomIntRef.current(2))}>
            {symbols.map((symbol, index) => {
                const basis = getRandomIntRef.current(100 / symbols.length);
                const offsetTop = 1 / (getRandomIntRef.current(100) + 1);

                return (
                    <AnimatedSymbol
                        key={`loading-${index}`}
                        basis={basis}
                        offsetTop={offsetTop}
                    >
                        {symbol}
                    </AnimatedSymbol>
                );
            })}
        </LoadingIndicatorWrapper>
    );
};

LoadingIndicator.displayName = "LoadingIndicator";

LoadingIndicator.defaultProps = {
    getRandomInt,
};

export default LoadingIndicator;
