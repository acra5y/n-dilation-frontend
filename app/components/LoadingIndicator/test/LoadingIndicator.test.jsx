import React from "react";
import { shallow, mount } from "enzyme";

import LoadingIndicator from "../LoadingIndicator";
import * as WindowContext from "../../WindowContext";
import * as AnimatedSymbol from "../AnimatedSymbol";
import { INTERVAL } from "../constants";

const amountOfSymbolsLimit = [3];
const symbolIndexLimit = [27];
const flexDirectionLimit = [2];
const flexBasisLimit = [100 / (amountOfSymbolsLimit[0] + 1)];
const offsetTopLimit = [100];

const getRandomIntCallArgsPerRenderCycle = [
    amountOfSymbolsLimit,
    symbolIndexLimit,
    symbolIndexLimit,
    symbolIndexLimit,
    symbolIndexLimit,
    flexDirectionLimit,
    flexBasisLimit,
    offsetTopLimit,
    flexBasisLimit,
    offsetTopLimit,
    flexBasisLimit,
    offsetTopLimit,
    flexBasisLimit,
    offsetTopLimit,
];

const createWindow = (
    setInterval = jest.fn((cb) => {
        cb();
        return "interval-id";
    }),
    clearInterval = jest.fn()
) => ({
    setInterval,
    clearInterval,
});

const createDefaultProps = () => ({
    getRandomInt: jest.fn().mockImplementation((arg) => arg),
});

const render = (
    props = createDefaultProps(),
    window,
    renderMethod = shallow
) => {
    jest.spyOn(WindowContext, "useWindowContext").mockImplementation(
        () => window
    );
    return renderMethod(<LoadingIndicator {...props} />);
};

describe("LoadingIndicator", () => {
    it("should render initially with a set of symbols", () => {
        const component = render();

        expect(component).toMatchSnapshot();
    });

    it("should use getRandomInt from props for randomization of symbols and their styles", () => {
        const props = createDefaultProps();
        render(props);

        expect(props.getRandomInt.mock.calls).toEqual(
            getRandomIntCallArgsPerRenderCycle
        );
    });

    it("should use getRandomInt from props for randomization of symbols and their styles 2s", () => {
        jest.spyOn(AnimatedSymbol, "default").mockImplementation(() => (
            <>mock-component</>
        ));
        const props = createDefaultProps();
        const window = createWindow();
        render(props, window, mount);

        const extraCallForInitialState = [
            amountOfSymbolsLimit,
            symbolIndexLimit,
            symbolIndexLimit,
            symbolIndexLimit,
            symbolIndexLimit,
        ];

        expect(window.setInterval.mock.calls.length).toEqual(1);
        expect(window.setInterval.mock.calls[0][1]).toEqual(INTERVAL);
        expect(props.getRandomInt.mock.calls).toEqual([
            ...getRandomIntCallArgsPerRenderCycle,
            ...extraCallForInitialState,
            ...getRandomIntCallArgsPerRenderCycle,
        ]);
    });
});
