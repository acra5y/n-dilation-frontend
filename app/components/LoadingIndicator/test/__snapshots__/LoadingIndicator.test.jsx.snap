// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicator should render initially with a set of symbols 1`] = `
<LoadingIndicator__LoadingIndicatorWrapper
  reverse={true}
>
  <AnimatedSymbol
    basis={32.333333333333336}
    getRandomInt={[Function]}
    key="loading-0"
    offsetTop={0.01}
  >
    j→∞
  </AnimatedSymbol>
  <AnimatedSymbol
    basis={32.333333333333336}
    getRandomInt={[Function]}
    key="loading-1"
    offsetTop={0.01}
  >
    j→∞
  </AnimatedSymbol>
  <AnimatedSymbol
    basis={32.333333333333336}
    getRandomInt={[Function]}
    key="loading-2"
    offsetTop={0.01}
  >
    j→∞
  </AnimatedSymbol>
</LoadingIndicator__LoadingIndicatorWrapper>
`;
